internal/servers/repository_workerauth.go

```diff
@@ -0,0 +1,729 @@
+		if err != nil {
+			return fmt.Errorf("(%s) Transaction rollback failed: %w", op, err)
+		}
+	} else {
+		err := r.workerAuthTransaction.transaction.Commit(nil)
+		if err != nil {
+			return fmt.Errorf("(%s) Transaction commit failed: %w", op, err)
+		}
+	}
+	return nil
+}
+
+// Store implements the Storage interface
+func (r *WorkerAuthRepositoryStorage) Store(ctx context.Context, msg nodee.MessageWithId) error {
+	const op = "workerauth.(WorkerAuthRepositoryStorage).Store"
+	if err := types.ValidateMessage(msg); err != nil {
+		return fmt.Errorf("(%s) given message cannot be stored: %w", op, err)
+	}
+	if msg.GetId() == "" {
+		return fmt.Errorf("(%s) given message cannot be stored as it has no ID.", op)
+	}
+
+	// Determine type of message to store
+	marshaledBytes, err := proto.Marshal(msg.(proto.Message))
+	if err != nil {
+		return fmt.Errorf("error marshaling nodee.MessageWithId: %w", err)
+	}
+	switch msg.(type) {
+	case *types.NodeInformation:
+		node, err := unmarshalNodeInformation(ctx, marshaledBytes)
+		if err != nil {
+			return err
+		}
+		err = r.storeNodeInformation(ctx, node)
+		if err != nil {
+			return errors.Wrap(ctx, err, op)
+		}
+	case *types.RootCertificate:
+		cert, err := unmarshalRootCertificate(ctx, marshaledBytes)
+		if err != nil {
+			return err
+		}
+		err = r.storeRootCertificate(ctx, cert)
+		if err != nil {
+			return errors.Wrap(ctx, err, op)
+		}
+	default:
+		err = fmt.Errorf("(%s) Message type not supported for Store", op)
+		return errors.Wrap(ctx, err, op)
+	}
+
+	return nil
+}
+
+// Node information is stored in two parts:
+// * the workerAuth record is stored with a reference to a worker
+// * certificate bundles are stored with a reference to the workerAuth record and issuing root certificate
+func (r *WorkerAuthRepositoryStorage) storeNodeInformation(ctx context.Context, node *types.NodeInformation) error {
+	const op = "workerauth.(WorkerAuthRepositoryStorage).storeNodeInformation"
+
+	if r.worker == nil {
+		return errors.Wrap(ctx, fmt.Errorf("Cannot store workerAuthentication without a worker reference"), op)
+	}
+
+	nodeAuth := AllocWorkerAuth()
+	nodeAuth.WorkerKeyIdentifier = node.Id
+	nodeAuth.WorkerEncryptionPubKey = node.EncryptionPublicKeyBytes
+	nodeAuth.WorkerSigningPubKey = node.CertificatePublicKeyPkix
+	nodeAuth.Nonce = node.RegistrationNonce
+
+	// Encrypt the private key
+	databaseWrapper, err := r.kms.GetWrapper(ctx, scope.Global.String(), kms.KeyPurposeDatabase)
+	if err != nil {
+		return errors.Wrap(ctx, err, op)
+	}
+	nodeAuth.KeyId, err = databaseWrapper.KeyId(ctx)
+	if err != nil {
+		return errors.Wrap(ctx, err, op)
+	}
+	nodeAuth.ControllerEncryptionPrivKey, err = encrypt(ctx, node.ServerEncryptionPrivateKeyBytes, databaseWrapper)
+	if err != nil {
+		return errors.Wrap(ctx, err, op)
+	}
+
+	nodeAuth.WorkerId = r.worker.PublicId
+
+	err = nodeAuth.ValidateNewWorkerAuth(ctx)
+	if err != nil {
+		return errors.Wrap(ctx, err, op)
+	}
+
+	onConflict := &db.OnConflict{
```

Comment:
Thinking through concurrent operations...

NodeInformation can be written either through the API or, based on a discussion today, possibly at registration request time (if we move back away from caching). If it's written via the API, e.g. through AuthorizeNode, then if a request comes before that to fetch credentials it's fine, and if it comes after that it's fine. If a node reaches out to multiple controllers though and we store a registration request we want to ensure we only generate one ID, so we will need to ensure that an upsert takes into account whether we have a worker with an ID for that key ID already. (I'm not entirely sure what the conflict is doing right below this.)

On the root certificate side if we had multiple controllers seeing no certificates and trying to create an initial set, we'd want to make sure only one is successful. That's probably pretty simple in a transaction, just checking if current and next already exist.

I think NodeInformation is a bit tricker though, we may need a Version or as a constraint in the transaction. So that if we read a NodeInformation object in a transaction, and we are trying to update it, we ensure that we are only updating it if the Version hasn't changed.